lib/mpl_toolkits/mplot3d/axes3d.py

```diff
@@ -1655,6 +1651,24 @@ def plot_surface(self, X, Y, Z, *args, norm=None, vmin=None,
                     if fcolors is not None:
                         colset.append(fcolors[rs][cs])
 
+        # In cases where there are NaNs in the data (possibly from masked
+        # arrays), artifacts can be introduced. Here check whether NaNs exist
+        # and remove the entries if so
+        if not isinstance(polys, np.ndarray) or np.isnan(polys).any():
+            new_polys = []
+            new_colset = []
+
+            for p, col in itertools.zip_longest(polys, colset):
```

Comment:
This is quite clever for handling the two `fcolor` cases. However, I think we should explain it in a comment.

> depending on fcolors, colset is either an empty list or has as many elements as polys.
> In the former case new_colset results in a list with None entries, that is discarded later.

Reply:
I've added your suggested comment locally, I'll wait for further changes (e.g. adding tests) before pushing to save running the CI